Ms 14 change password

DamianBednarek opened the pull request:
- implemented change password functionality
- wrote API documentation
- wrote tests for changing password

Commit:
Initial commit

Commit:
Profile View and Edit

Implemented profile viewing and editing.
tested the implementation

Commit:
API Documentation

DamianBednarek committed:
MS-14-change-password

- implemented change password functionality
- wrote API documentation
- wrote tests for changing password

DamianBednarek committed:
Update custom_user_schema.py

dmitry-sergeev commented on MusicHub/users/serializers.py:
```diff
@@ -1,3 +1,4 @@
+from dataclasses import fields
```
unused import

DamianBednarek replied:
Removed unused import

dmitry-sergeev commented on MusicHub/users/serializers.py:
```diff
@@ -68,6 +69,36 @@ class Meta:
+        fields = ["password", "confirm_password", "old_password"]
+
+    def validate(self, attrs):
+        if not attrs["password"] == attrs["confirm_password"]:
```
Does it make sense to move to some service like `ProfileService`?

DamianBednarek replied:
I've created a separate _profile_service.py_ file to store passwords validation logic

dmitry-sergeev commented on MusicHub/users/test/test_change_password.py:
```diff
@@ -0,0 +1,100 @@
+        self.user_data = UserFactory()
+        self.user_data.set_password("abcABC123*")
+        self.user_data.save()
+        self.client.force_authenticate(user=self.user_data)
```
All staff into `setUp` will be duplicates for all tests where we need to have authorized user. So some base class for the tests like `AuthorizedApiTestCase` or make decorator which can be easy shared for the classes and test functions

DamianBednarek replied:
I've created a base class for unauthorized user test case and authorized user test case

dmitry-sergeev commented on MusicHub/users/test/test_change_password.py:
```diff
@@ -0,0 +1,100 @@
+        response = self.client.patch(self.url, data=data)
+        self.assertEqual(response.status_code, 400)
+        user = User.objects.get(email=self.user_data.email)
+        self.assertFalse(user.check_password("password123"))
```
the expected result is the same for most tests
```
user = User.objects.get(email=self.user_data.email)
self.assertFalse(user.check_password("password123"))
```
think about to not duplicate it (it can be separate func, decorator, ...)

DamianBednarek replied:
Moved this logic to base test case classes

dmitry-sergeev commented on MusicHub/users/test/test_profile.py:
```diff
@@ -0,0 +1,57 @@
+        self.assertContains(get_response, "verified@email.com")
+        self.assertContains(get_response, "verified")
+        self.assertContains(get_response, "user")
+        self.assertNotContains(get_response, "password")
```
you can use smth like 
```
EXPECTED_RESPONSE_KEYS = ['email', 'first_name', 'last_name', ....]
```

Henok-U replied:
updated test as per comment

dmitry-sergeev commented on MusicHub/users/views/profile_view.py:
```diff
@@ -0,0 +1,79 @@
+from MusicHub.users.serializers import ChangePasswordSerializer, ProfileSerializer
+
+
+class ProfileView(GenericAPIView):
```
I remember that you didn't use more specific views because problems with swagger but we need to resolve it in order to be free use DRF

Henok-U replied:
updated view using a specific view.

DamianBednarek committed:
Fixed pointed problems in pull request

- Created Profile service for validating passwords in user serializer
- Created base class for Tests, including unauthorized user and authorized
- Fixed saving password in user factory
- Refactoring imports

DamianBednarek committed:
Update utils.py

DamianBednarek committed:
Merge branch 'dev' into MS-14-change-password
